{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorial to estimate the Geographic Spectrum of Shared Alleles (GSSA)\n",
    "\n",
    "This is a tutorial for estimating the GSSA and use it to uncover the spatial dynamics of potential range expansions. You can run the ipython notebook that created this step-by-step estimation at *notebook/GSSA_calculation.ipynb*.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## STEP 1. Estimating the geo-genetic and geographic-distance histograms\n",
    "\n",
    "First import the necessary packages:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys, subprocess, warnings, math, itertools as it, numpy as np\n",
    "from collections import Counter\n",
    "from scipy import spatial\n",
    "import plotly, plotly.plotly as py, matplotlib.pyplot as plt, plotly.tools as tls, plotly.graph_objs as go\n",
    "plotly.tools.set_credentials_file(username='fern81', api_key='F5a4CnfP6WhcwvPlUYK8')\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, define the path to the two input files necessary for this step:\n",
    "\n",
    "- file containing a *mSFS* file describing the joint site frequency spectrum for all sampled localities (see associated bitbucket repository for details on how to obtain this file from a vcf listing the sampled SNPs)\n",
    "\n",
    "- file listing either the geographic coordinates of sampled localities (one line per sampled locality) or the pairwise \n",
    "  effective distance between them [**note:** for the current example, we would use a geographic coordinates files]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "infilename = '../example_files/multiSFS_Sampled_SNPs.sfs'\n",
    "geoname = '../example_files/Sampled_locs.coord'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the information contained in these two input files, the calculation proceeds by:\n",
    "\n",
    "- defining the output file's name:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "if '/' in infilename:\n",
    "    outdir = infilename.split('/')\n",
    "    outdir = '/'.join(outdir[:-1])\n",
    "    infile = infilename.split('/')[-1]\n",
    "else:\n",
    "    outdir = './'\n",
    "    infile = infilename\n",
    "    \n",
    "prefix = infile.replace('multiSFS_','').replace('.sfs','')\n",
    "outfilename = '%s/%s.histo' %(outdir, prefix)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- reading the sampled localities' geographic file, extracting all pairwise geographic distances between them, and storing those distances in an *array* (**note:** the grographic file could either list the geographic coordinates of all sampled sites (file suffix \"*.coord*\"), or provide the pairwise distances between them )file siffix \"*.dist*\"):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "if geoname.endswith('coord'):\n",
    "    #reading in location coordinates and calculating geographic distance among them\n",
    "    coords = np.empty([1,3])\n",
    "    nloc = -1       #stores number of localities sampled\n",
    "    with open(geoname, 'r') as coordfile:\n",
    "        for line in coordfile:\n",
    "            nloc += 1\n",
    "            if nloc > 0:\n",
    "                row = line.strip().split()\n",
    "                row[0] = row[0].replace('P','')\n",
    "                coords = np.vstack((coords, [float(x) for x in row]))\n",
    "    coords = np.delete(coords,0,0)\n",
    "    EuD = spatial.distance.pdist(coords[:,1:],'euclidean')\n",
    "    AllEuD = list(EuD)\n",
    "    popnames = coords[:,0]\n",
    "    \n",
    "elif geoname.endswith('.dist'):\n",
    "    #reading in pairwise distances between locations\n",
    "    with open(geoname, 'r') as distfile:\n",
    "        n = 0\n",
    "        for line in distfile:\n",
    "            n += 1\n",
    "            if n == 1:\n",
    "                popnames = [float(x.replace('P','')) for x in line.strip().split(',')]\n",
    "                nloc = len(popnames)\n",
    "            else:\n",
    "                break\n",
    "    distmat = np.genfromtxt(geoname, delimiter=',')\n",
    "    distmat = np.delete(distmat,0,0)\n",
    "    AllEuD = []\n",
    "    n = -1\n",
    "    for r in xrange(np.shape(distmat)[0]):\n",
    "        n += 1\n",
    "        for c in xrange(np.shape(distmat)[1]):\n",
    "            if c > n and r!=c:\n",
    "                AllEuD.append(distmat[r,c])\n",
    "    EuD = np.array(AllEuD)\n",
    "\n",
    "else:\n",
    "    #stop with an error if the incorrect files are provided as input\n",
    "    sys.exist('Geography file should be either of type \".coord\" or \".dist\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- creating a *list* with the definition of the entries on the *mSFS*, which indicate the set of combinations of minor alleles joint frequencies (see [fastsimcoal2 manual](http://cmpg.unibe.ch/software/fastsimcoal2/man/fastsimcoal26.pdf) for details). **NOTE:** currently this calculation assumes only 1 diploid individual per locality, and biallelic SNPs, though the script could be easily expanded to relax the first assumption."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "combs = []\n",
    "for combi in it.product(range(3), repeat = nloc):\n",
    "    combs.append(combi)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- labeling distance pairs in the pairwise distance array:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "comb = []\n",
    "for i in it.combinations(popnames, 2):\n",
    "    comb.append(str(int(i[0])) + '/' + str(int(i[1])))\n",
    "\n",
    "EuD = np.vstack((comb,EuD))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- creating a common binning scheme for constructing both histograms using Sturge's (1926) rule:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sturges(vals):\n",
    "    N = len(vals)\n",
    "    nbins = math.ceil(math.log(N,2) + 1)\n",
    "    bins = np.linspace(min(vals), max(vals), nbins)\n",
    "    return list(bins)\n",
    "\n",
    "bins = sturges(AllEuD)\n",
    "bins.insert(0, 0.0)\n",
    "#a -1 is added below to the binning scheme to temporally store\n",
    "#the singleton SNPs, which are not used for the GSSA creation\n",
    "#as they are more prone to correspond to call errors\n",
    "bins.insert(0, -1.0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- creating a null histogram expectation exclusively based on samples' relative position to each other (i.e., generating the geographic-distance histogram)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "geobins = {}\n",
    "for l in list(popnames):\n",
    "    loc = str(int(l))\n",
    "    dists = [0]     #distance of each locality with itself, that is a distance of zero\n",
    "    for x in xrange(len(comb)):\n",
    "        if comb[x].startswith(loc+'/') or comb[x].endswith('/'+loc):\n",
    "            dists.append(AllEuD[x])\n",
    "    geo_hist = list(np.histogram(dists, bins)[0])\n",
    "    geobins.setdefault(loc, geo_hist[1:])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-plotting the geographic histograms just created:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plotting geographic-distance histograms\n",
    "if np.shape(coords)[0] % 2 == 0:\n",
    "    dim = np.shape(coords)[0] / 2\n",
    "else:\n",
    "    dim = np.shape(coords)[0] / 2 + 1\n",
    "xvals = range(1,len(geo_hist))\n",
    "mpl_fig, axs = plt.subplots(dim,2,figsize=(10,5))\n",
    "mpl_fig.suptitle('Geographic_histograms', fontsize=16)\n",
    "mpl_fig.text(0.04, 0.5, 'Frequency',va='center', rotation='vertical', fontsize=14)\n",
    "n = 0\n",
    "row = -1\n",
    "keys = [int(l) for l in geobins.keys()][:]\n",
    "keys.sort()\n",
    "for k in keys:\n",
    "    n += 1\n",
    "    if n%2==1:\n",
    "        row += 1\n",
    "        col = 0\n",
    "    elif n%2==0:\n",
    "        col = 1\n",
    "    name = 'Loc%s' %k\n",
    "    axs[row,col].axes.bar(xvals, geobins[str(k)], align='center', alpha=0.5, color='SkyBlue', label=name)\n",
    "    axs[row,col].xaxis.set_visible(False)\n",
    "    axs[row,col].legend(loc=\"upper right\")\n",
    "plotly_fig = tls.mpl_to_plotly(mpl_fig)\n",
    "figpath = outdir + '/geographic histogram'\n",
    "py.iplot(plotly_fig, filename=figpath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- reading *mSFS* file in and generating the geo-genetic histogram (called genHist below):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(outfilename, 'w') as outfile:\n",
    "    #print header to output file storing geographic-distance and geo-genetic histograms\n",
    "    print >> outfile, 'Pop\\t%s\\t%s' %('\\t'.join(['geoB'+str(x) for x in range(1,(len(bins)-1))]), '\\t'.join(['genB'+str(x) for x in range((len(bins)-1))]))\n",
    "    \n",
    "    #read in mSFS file\n",
    "    pathF = infilename\n",
    "    n = 0\n",
    "    with open(pathF, 'r') as sfsfile:\n",
    "        genHist = {}\n",
    "        for line in sfsfile:\n",
    "            n += 1\n",
    "            \n",
    "            #setting a dictionary for each locality to store the aggregated relative spatial distribution of minor alleles (named as vector \"Si\" in the publication)\n",
    "            if n == 1:\n",
    "                pops = [str(int(x.replace('\"P','').replace('\"',''))) for x in line.strip().split()[-nloc:]]   #(int(p)-1 because dadi starts pops at 1, but populations in coordinate file are labeled starting at 0\n",
    "                for p in pops:\n",
    "                    genHist.setdefault(p,[])\n",
    "                if set(pops) != set([str(int(x)) for x in popnames]):\n",
    "                    os.remove(outfilename)\n",
    "                    sys.exit('Mismatched sampled localities and coordinates!')\n",
    "    \n",
    "            #estimating the aggregated relative spatial distribution of minor alleles for each individual (named as vector \"Si\" in the publication)\n",
    "            if n == 2:\n",
    "                sfs = [float(x) for x in line.strip().split()]\n",
    "                for xbin in xrange(len(sfs)):\n",
    "                    if sfs[xbin] > 0.0:\n",
    "                        #the next two lines check for splitted bins (i.e., non-integer) that are present in unfolded SFS\n",
    "                        if not float(sfs[xbin]).is_integer():\n",
    "                            os.remove(outfilename)\n",
    "                            sys.exit('Unfolded SFS provided, folded required!')\n",
    "                        #the next two lines check whether the combined frequency of each SNP is smaller than that of the alternative allele, which should be the case when using the minor allele for polarization\n",
    "                        if sum(list(combs[xbin])) > nloc:\n",
    "                            warnings.warn('Input SFS is not polarized based on minor allele')\n",
    "                        \n",
    "                        #storing the distance between copies of the minor alles for each SNP locus and individual\n",
    "                        for j in xrange(len(combs[xbin])):\n",
    "                            if combs[xbin][j] != 0:\n",
    "                                pop = pops[j]\n",
    "                                popsIN = [x for x in EuD[0,] if x.startswith(pop+'/') or x.endswith('/'+pop)]\n",
    "                                popsTO = [x.replace(pop+'/','').replace('/'+pop,'') for x in popsIN]\n",
    "                                indexes = [pops.index(x) for x in popsTO]\n",
    "                                indexPop = pops.index(pop)\n",
    "                                distpairs = []\n",
    "                                count = Counter(list(combs[xbin]))\n",
    "                                if count.get(0)==(nloc-1):\n",
    "                                    for freq in xrange(int(sfs[xbin])):\n",
    "                                        distpairs.append(-1.0)  #special value for private SNPs at frequency of 1. Because such allele may be an error, it is not actually consider in the ragedness index calculation, but here such instances are recorded just in case\n",
    "                                else:\n",
    "                                    if combs[xbin][j] == 2:\n",
    "                                        for freq in xrange(int(sfs[xbin])):\n",
    "                                            distpairs.append(0.0)\n",
    "                                    TOcombs = np.array([pops, list(combs[xbin])])\n",
    "                                    TOcombs = np.delete(TOcombs,indexPop,1)\n",
    "                                    for k in xrange(len(popsIN)):\n",
    "                                        ToPop = popsIN[k].split('/')\n",
    "                                        del ToPop[ToPop.index(pop)]\n",
    "                                        indexToPop = list(TOcombs[0,]).index(ToPop[0])\n",
    "                                        if TOcombs[1,indexToPop] != '0':\n",
    "                                            dist = float(EuD[1,list(EuD[0,]).index(popsIN[k])])\n",
    "                                            times = int(TOcombs[1,indexToPop]) * combs[xbin][j]\n",
    "                                            for t in xrange(times):\n",
    "                                                for freq in xrange(int(sfs[xbin])):\n",
    "                                                    distpairs.append(dist)\n",
    "                                genHist[pop].extend(distpairs)\n",
    "    \n",
    "    #writing out geographic-distance and geo-genetic histograms to file\n",
    "    for loc in [str(int(x)) for x in popnames]:\n",
    "        print >> outfile, 'P%s\\t%s\\t%s' %(loc, '\\t'.join([str(x) for x in geobins.get(loc)]), '\\t'.join([str(y) for y in list(np.histogram(genHist.get(loc), bins)[0])]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-plotting the geo-genetic histograms just created:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plotting geo-genetic histograms\n",
    "list(np.histogram(genHist.get(loc), bins)[0])\n",
    "mpl_fig, axs = plt.subplots(dim,2,figsize=(10,5))\n",
    "mpl_fig.suptitle('Geo-genetic_histograms', fontsize=16)\n",
    "mpl_fig.text(0.04, 0.5, 'Frequency', va='center', rotation='vertical', fontsize=14)\n",
    "n = 0\n",
    "row = -1\n",
    "for k in keys:\n",
    "    n += 1\n",
    "    if n%2==1:\n",
    "        row += 1\n",
    "        col = 0\n",
    "    elif n%2==0:\n",
    "        col = 1\n",
    "    genBins = list(np.histogram(genHist.get(str(k)), bins)[0])[1:]\n",
    "    name = 'Loc%s' %k\n",
    "    axs[row,col].axes.bar(xvals, genBins, align='center', alpha=0.5, color='SkyBlue', label=name)\n",
    "    axs[row,col].xaxis.set_visible(False)\n",
    "    axs[row,col].legend(loc=\"upper right\")\n",
    "plotly_fig = tls.mpl_to_plotly(mpl_fig)\n",
    "figpath = outdir + '/geogenetic_histogram'\n",
    "py.iplot(plotly_fig, filename=figpath)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- printing output file name to be used in **next step**, which can be run following the other ipython notebook store in the *notebook/* directory named *GSSA_calculation_step2.ipynb*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../example_files/Sampled_SNPs.histo\n"
     ]
    }
   ],
   "source": [
    "print outfilename"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## STEP 2. Estimating the GSSA and inferring the sampled locality closest to the expansion source\n",
    "\n",
    "First load *R*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython\n",
    "\n",
    "import warnings \n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, import the necessary *R* packages:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "suppressMessages(require(matrixStats))\n",
    "require(MASS)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, define the path to the two input files necessary for this second step:\n",
    "\n",
    "- file generated in **Step 1** containing the geographic-distance and geo-genetic histograms (file suffix \"*.histo*\")\n",
    "\n",
    "- file listing the geographic information of sampled localities; either the geographic coordinates (file suffix \"*.coord*\") or the pairwise distances between localities (file suffix \"*.dist*\") [**note:** for the current example, we would use a geographic coordinates files]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "infile = '../example_files/Sampled_SNPs.histo'\n",
    "geofile = '../example_files/Sampled_locs.coord'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The calculation then proceeds by:\n",
    "\n",
    "- setting output file names:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "File.prefix = sub('.histo', '', basename(infile))\n",
    "File.path = dirname(infile)\n",
    "resultsFile = sprintf('%s/%s_GSSA_results.tsv', File.path, File.prefix)\n",
    "raggFile = sprintf('%s/%s_raggedness_index.tsv', File.path, File.prefix) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- reading the geographic file to get the number and names of sampled localities:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "if (endsWith(geofile, '.coord')){\n",
    "    coord = read.table(geofile, header=T, sep='\\t')\n",
    "    dists = as.matrix(dist(coord[,2:3]))\n",
    "    colnames(dists) = coord$Pop\n",
    "} else {\n",
    "    if (endsWith(geofile, '.dist')){\n",
    "        dists = read.csv(geofile, header=T)\n",
    "        dists = as.matrix(dists, ncol=ncol(dists))\n",
    "    }\n",
    "}\n",
    "rownames(dists) = colnames(dists)\n",
    "popnames = colnames(dists)\n",
    "numloc = length(popnames)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- reading histograms input file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "d = read.table(infile, header=TRUE, sep='\\t') "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- writing output files header:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "write(paste('Source', 'Longitude', 'Latitude', sep='\\t'), file=resultsFile, append=F)\n",
    "write(paste(paste(d$Pop[1:numloc], collapse='\\t'), sep='\\t'), file=raggFile, append=F)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- defining function to calculate the GSSA statistic:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "GSSA.func = function(data){\n",
    "    #data = dataframe storing the geographic-distance and geo-genetic distances\n",
    "    \n",
    "    rownames(data) = data$Pop\n",
    "    geoBins = data[grep('geoB',names(data))]\n",
    "    genBins = data[grep('genB',names(data))[-1]]   \n",
    "    #the [-1] above is to remove the singleton category bin, which is not used in the GSSA calculation geoBins.corrected = genBins \\\n",
    "    genBins.corrected = genBins\n",
    "    genBins.corrected[which(!is.na(genBins.corrected), arr.ind=TRUE)] = NA\n",
    "    reg = lm(as.numeric(genBins) ~ as.numeric(geoBins))\n",
    "    for (i in 1:length(genBins.corrected)){\n",
    "        genBins.corrected[i] = abs(reg$residuals[i])\n",
    "    }\n",
    "    return(genBins.corrected)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- defining function to plot GSSA vectors for each sampled locality:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "plot.gssa = function(x, Peaks){\n",
    "    #x = dataframe row storing the corrected geo-genetic distances (i.e., GSSA)\n",
    "    #Peaks = vector of estimated raggedness indexes\n",
    "    \n",
    "    if((length(Peaks) %% 2) == 0) {\n",
    "        rows = length(Peaks) / 2\n",
    "    } else {\n",
    "        rows = (length(Peaks) / 2) + 1\n",
    "    }\n",
    "    par(mfrow=c(rows,2))\n",
    "    \n",
    "    for (i in 1:nrow(x)){\n",
    "        x = x[order(Peaks),]\n",
    "        barplot(as.numeric(x[i,]), ylab='Frequency', main=sprintf('GSSA: %s [Rank: %d]', rownames(x)[i], i))\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- defining a function to calculate Harpending's (1994) raggedness index:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "diff.func = function(x){\n",
    "     #x = dataframe row storing the geographic-distance and corrected geo-genetic distances (i.e., GSSA)\n",
    "\n",
    "     require(MASS)\n",
    "     null = x[grep('geoB',names(x))]\n",
    "     sfs = x[grep('genB',names(x))]\n",
    "     non0bins = as.numeric(which(null!=0.0))\n",
    "     sfs = sfs[non0bins]\n",
    "     comb = combn(seq(length(sfs)),2)\n",
    "     paired = combn(sfs,2)\n",
    "     paired = paired[,which(colDiffs(comb)==1)]\n",
    "     paired = matrix(as.numeric(paired), nrow=2)\n",
    "     diffs2 = (colDiffs(paired))^2\n",
    "     return((sum(diffs2)/(length(non0bins)-1)))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- putting together all functions described above into a wrapper function to identify the sampled location closest to the expansion source, plot the GSSA vectors, and write results to outpu files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "ragg.func = function(data, numloc){\n",
    "    #data = dataframe storing the geographic-distance and geo-genetic distances\n",
    "    #numloc = number of sampled localities \n",
    "    \n",
    "    #estimating the GSSA vectors for each sampled population\n",
    "    #gssa = apply(data, 1, GSSA.func) --> for an unknow reason apply fails here in the notebook, so:\n",
    "    gssa = data.frame()\n",
    "    for (i in 1:nrow(data)){\n",
    "        gssa = rbind(gssa, GSSA.func(data[i,]))\n",
    "    }\n",
    "    gssa = as.data.frame(gssa)\n",
    "    \n",
    "    #calculating Harpending's (1994) raggedness index on GSSA vectors\n",
    "    #Peaks = apply(cbind(geoBins = data[grep('geoB',names(data))], gssa), 1, diff.func)  --> for an unknow reason apply fails here in the notebook, so:\n",
    "    Peaks = rep(NA,numloc)\n",
    "    names(Peaks) = popnames\n",
    "    for (i in 1:nrow(data)){\n",
    "        Peaks[i] = diff.func(cbind(geoBins = data[i,grep('geoB',names(data))], gssa[1,]))\n",
    "    } \n",
    "    \n",
    "    #identify sampled location closest to range expansion source\n",
    "    source = names(Peaks)[which.min(Peaks)]\n",
    "    if (endsWith(geofile, 'coordfile')){\n",
    "        coord.source = coord[which(coord$Pop == source), c('longitude', 'latitude')]\n",
    "    } else {\n",
    "        cat('\\nWARNING!: coordinate file not provided, potential source coordinates reported as NA\\n\\n')\n",
    "        coord.source = c(NA, NA)\n",
    "    }\n",
    "    \n",
    "    #write out results\n",
    "    write(paste(Peaks, collapse='\\t'), file=raggFile, append=T)\n",
    "    write(paste(source, as.numeric(coord.source[1]), as.numeric(coord.source[2]), sep='\\t'), file=resultsFile, append=T)\n",
    "    \n",
    "    #plot GSSA vectors (ordered by inferred time of colonization)\n",
    "    plot.gssa(gssa, Peaks)\n",
    "   }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- estimating GSSA vectors and sampled lcoation closest to the range expansion source and plotting the GSSA vectors, ranked by estimated time of colonization:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "WARNING!: coordinate file not provided, potential source coordinates reported as NA\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeAAAAHgCAMAAABKCk6nAAAC3FBMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8RERESEhITExMUFBQVFRUWFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+hoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+xsbGysrKzs7O1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjLy8vMzMzNzc3Ozs7Pz8/R0dHS0tLT09PU1NTV1dXW1tbX19fY2Nja2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7///9/l72tAAAd4UlEQVR4nO2djV9UZb7Ah2odtS4k6q6ie/M9cxVDM8BVYRA3KNxENEokM5DcNLLI0rXUFMTYDJ3C0TXAO8q2hVetdbM0DW23C8g14xILvqwpoMiLvPz+gXvmeQYNzpkzMwTnHH/P7/v5dHg4c2Z8vvOdM4dzmE+YgECNSe8JEL0LBUYOBUYOBUYOBUYOBUYOBUYOBUYOBUYOBUYOBUYOBUYOBUYOBUaOMQJv7jvyhmyl3dz5+1DfxRpNp9fR0LfXAtsn9R+YVF8XN9Av8BRfAiyftK3j5jyT6a7RhTce6/fwd9J3m0McK0y+cfU/fYTbwvVpv3gDINnIgT3y5R7ufa+E9xv1Wc/49lbg4z5ZF4+Nzc8117SssfElNAw9PQHgoukLcAg3XksZZP31lScXglO4sfGY6dDlGSPGF+cNfnFIbLvd3D4vpppt/TfrFIMH9syXe7j33T79fGy4sQOnjWNfyocHr/20jS9hzwKYXATNh6+AQ1jaxn9ZDGQ9CE5hf39zYtM3Ge2zltlNxadMZ+zmVSENfGuAaQYP7Jkv9/DEt3rURmMHXj0a6s19pkHLyawxMc5lsPSu9KzzdsnP13fvH56Ad8aDU7jqqM8ROB0RHfC0vQ/8aCqymwZE3Ho8owf2zLdTYDXfr4a83mbswF/7pF/6IXGqdU5106qhfFl6Ty1827+u4xVddf4m7A7499xF4BQGmD+xLTayNTBOOho5hM2X7s+/U/Zgz3w7BVbxPdl33/Ubxg4M+yf2848tvfC4X98JhXy5Yoa0OsDacUxybNQYd9+0CrglfPZuW/6AOVn3JTuFIXtwGdt61713/eLeViMH9syXe7j3nS/t+780eGCvkIQ9wNCBvUJDX4MEVjov7Ar28+CuGPw8mDAGPPBjtjqd56EtAvnywA37YhO+0HkmWiKQrzPwh5b4jJd1noqGCOTLA0fsqAHI0XkqGiKQLw9sryjZr/NENEUgXx54SVvbczpPRFME8uWBl1aWoznH9ASBfHng4pSUYp0noikC+fLAK6KionSeiKYI5MsDZ+k8C60RyJcHXnP4C0HO+zkC+fLAWzIyM3WeiKYI5MsDr51X9abOE9EUgXydb9HrINWTrb/MUeBMr06wVxDIlwd+6ZWSeJebVJSXl/8v//Vl9Op1MhLe0mSiPYqab+vnR44c+Ws1GyPw5YHLXlhe5mqL2uSkpKTpf2Lj6INHZbxzZwkz1HyvZ6Snp/8+m40R+PLAUVFhYa62qImbP3/+w1vYGIEwQ823drbFYhmXwcZe+xYpvaf/s6en7w23PtHh8ncrrQc++eSTZXvY2GvhD+cr8FHPTb/7uPRt/qigoGCpjY299l28Uv6evmxlT8/dG3jgoqITia62uP5GWlpa5Hts7L3wtr0yNqkJ5wQpoHZKc+UbBSpV7uBAzffGTmmni7eysfe+dvkdbGq+H7jwbf2tRU5IlRsvBXjgDRs2fdXlhvaadj6o+TlvWV4Lp+bI77BvkcodNoQ9JWN+tBtrNV9P3qLZi6Are9p70Ld5inz90bklbrwU4IFDQx8N7Xx1dkdMQswuNmrcLU3+6Z2qwi1/PyLnOLgWrhml8MIddbUbgddvld/hkLsLzaq+bA/+QNW3dKr8nXjdyCaVwD+Uy3F8IlyjwNJ7wsbO6x2TWsGFdzkC56gKnxn/vJxhKsIXwxTmH37etXDLMIVXxPDybgZW8236i+MYrP6CLp2rMP2pKr7/HhotJ+BiNwL/adgYGaMXugv8TGXlgs7rF33f+F0CG7G3rMDdbJwWLj8wzMwFqA5ROGJMbwXYFCZfP0s6BalTukNIrbQnzZKvD3tbOiZNd3FMypspXx/+qpvAAvnywCcSlxzvvP5catyqCjdP052LQL50LRo53l2LxoJAvh5ci0aIQL7ur0VjRCBfFrj9c72noS0i+fI9OOWAMB9hYQjk6/zITqYwH2FhCOTLAh+ErldmcSOSLwu8EnT9jZbmiORLgZHDAltsFptN75loiEi+LLDdgd4z0RCRfOl/woIcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcCowcYwTe3HfkDdlKu7nz96G+izWaTq+joW+vBbZP6j8wqb4ubqBf4Cm+BFg+aVvHzXkm012jCwGyTHnSd5tDHCtMvnH1P32E28JnHzE/8DEkGzmwR74FPn5+SeDelz8vPeLbW4GP+2RdPDY2P9dc07LGxpfQMPT0BICLJsdflcszNV5LGQRlvxl2K3Bj4zHTocszRowvzhv84pDYdru5fV5MNdv64ParkU8YOrBnvjvH8K3d+fLnxdCB08axL+XDg9d+2saXsGcBTC6C5sNXwCEsbePfMvXYf94K7O9vTmz6JqN91jK7qfiU6YzdvCqkgW8NF3yHlxg6sGe+WwZNeOgAuPflz4uhA68eDfXmPtOg5WTWmBjnMlh6V3rWebvk5+u7948vwe3AVUd9jsDpiOiAp+194EdTkd00IKLj4VrLF4YaOrBnvkXbS5cOBPe+/HkxdOCvfdIv/ZA41TqnumnVUL4svacWvu1f1/GKrjp/E8b7+vn0L3IKA8yf2BYb2RoYJx2NHMLmS/fn862TQyqWjjZ0YM9889Mup/4K3Pvy58XQgWH/xH7+saUXHvfrO6GQL1fMkFYHWDuOSc7tbu/B0s9Sd9vyB8zJui/ZKQzZg8vY1ueC+446ZOjAnvmenWIe+xm49wUw/B7sHZKwBxg6sFdo6GuQwErnhV3Bfh7cFYOfBxPGgAd+zFan8zy0RSBfHrhhX2yCKH/U3oFAvs7AH1riM17WeSoaIpAvDxyxowYgR+epaIhAvjywvaJkv84T0RSBfHngJW1tz+k8EU0RyJcHXlpZjuYc0xME8uWBi1NSinWeiKYI5MsDr4iKitJ5IpoikC8PnKXzLLRGIF8eeM3hLwQ57+cI5MsDb8nIzNR5IpoikC8PvHZe1Zs6T0RTBPJ1vkWvg1RPtv4yR4EzvTrBXkEgXx74pVdK4l1tUTPbYrEE/pmNo1evk5HwljYz7UnUfGuDg4KCxuaycY/5tv61QM5fW7s7fy/ggcteWF6mupl1H/sSffCojHfuwMBqvvUZ6enpc7PZuMd8L459Wc64892dvxfwwFFRYWGutmCv6Ae2sDGWwGq+N/Nzc3MX72TjngscJn+go+HaBQZ3v1tBtgeDim/dsqSkpN/yM2U0gYuKTiS62qLWcQwel8HGWAJr7qt34A0bNn3V5Yb2mvaffIdsD1bzrXstLS1t9rtsjCZwaOijoZ2vzu6ISYjZxUa1kdIr+sEePgZfHRYkZ9gVgFyLAlaAtscV1v/uEsAnSndId/PPq/ne3Gm1Wp9CdgzOBNjYef1K6b8Vt791swf/33CFXgEtrv9Zl8KpfyqUsXsRQHOgfH1hVAnA+g3y9QXufpOg5nvtdWkPjnxP1bf1H9/I+R+Vf9DrwC3+Cs/o4HMAVxT+5W/+5fqBeOBnKisXdF6/6PvG7xLYiB2TAnezcVq4fG+ZKZ0yVoco7EbTpdO8LIVpBkl7R53SHUJqpT1plnx92NvSUzpd6Q5VAHkz5evDX1V5rnvC9/wUBa1gyXenwiwt/6Xia1e6wwcqvvlKz2iau8AnEpcc77z+XGrcqgo3T9Odi0C+dC0aOd5di8aCQL7ur0VjRCBfD69FI0MgXxa4/XO9p6EtIvnyPTjlgDAfYWEI5Ov8yE6mMB9hYQjkywIfhK5XZnEjki8LvJJdqRMHkXwpMHJYYIvNYrPpPRMNEcmXBbY70HsmGiKSL/1PWJBDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFDgZFjjMCb+468IVtpN3f+PtR3sUbT6XU09O21wPZJ/Qcm1dfFDfQLPMWXAMsnbeu4Oc9kumt04ZXwfqM+k77bHOJYYfKNq//pI9wWLvDx80uCZCMH9sg3WZIcCe59W573+/Xfe8a3twIf98m6eGxsfq65pmWNjS+hYejpCQAXTY6/KpdnaryWMmj79POx4eAUbmw8Zjp0ecaI8cV5g18cEttuN7fPi6lmW+8c43hMIwf2zLfp+vWoNHDvmzvswidHjB04bRz7Uj48eO2nbXwJexbA5CJoPnwFHMLSNv4A1aM2glPY39+c2PRNRvusZXZT8SnTGbt5VUgD33rLoAkPHTB0YE99d41qAve+L0x6aPgnxg68ejTUm/tMg5aTWWNinMtg6V3pWeftkp+v7174asjrbeAUrjrqcwROR0QHPG3vAz+aiuymARHOrYu2ly4daOjAHvo2+O9zfOfO97lRdStGGjvw1z7pl35InGqdU920aihflt5TC9/2r+t4RVedvwkn++677vhpgwkDzJ/YFhvZGhgnHY0cwuZL9+fzrfPTLqf+ytCBPfOFPX7Njq3d+b4//PKyh4wdGPZP7OcfW3rhcb++Ewr5csUMaXWAteOY5NhovvQa/yXcEj57ty1/wJys+5KdwpA9uIxtfXaKeexnhg7smS+8OI19cefb9PR/jPzc4IG9QhL2AEMH9goNfQ0SWOm8sCvYz4O7YvDzYMIY8MCP2ep0noe2COTLAzfsi00Q5Y/aOxDI1xn4Q0t8xss6T0VDBPLlgSN21ADk6DwVDRHIlwe2V5Ts13kimiKQLw+8pK3tOZ0noikC+fLASyvL0ZxjeoJAvjxwcUpKsc4T0RSBfHngFVFRUTpPRFME8uWBs3SehdYI5MsDrzn8hSDn/RyBfHngLRmZmTpPRFME8uWB186relPniWiKQL7Ot+h1kNorD1+Uo8A/e+Wf8gZPfb9Umv6Z3p5dj8IDv/RKSbyrLWpnWyyWh/PY+LVwi4yZeSoPv3jlOhnLVvaoQXdQ9Y2UpIJy2Th6tXz6CW9pNs2egAcue2F5mastbmxNT0//vZWNow8elfGOmvBiu/wONv0Dq/lez5B852azsde+xoMHjooKC3O1xc33s7OzF+5kY1fCzX8pkFMIhg2s6ptjtVqf4r+JQBMYVH63UpuclJQUupWNXQmXBr4s5z+bXAduzkyXk9msgXAHLn3rlkm+0/mZMprARUUnEl1tURscFBT0wBY2dhl4rnz90akqgS8GZsqZfF4bZze+jp85xmWo+rpk13wF7NINleVyKh3rjyhwrgdVeeANGzZ91eWG9pp2Prj2Wlpa2ux3VYW9DxymcIdwKfD3Cu/1BS6Pl91FzbfOA1+XLN5VKGOb5Ht5aLScgEsAqU8+LyN+kTSdP6bJea1WegmFyH/ODV/uekY8cGjoo6Gdr87uiEmI2cVGN/fm5eUl2lSFey6wK+GWYUFyhpcDZD8oXz/Z3a+KVH13Oo7B6j9zuA7stW+OfP0+ybd5/AdyZpYArN8qv8MhlQvrPHAmwMbO6x0/CK1go2urHa/obVoFdiU8ReEOc7sh7N73dck38j1V3x9GK7zcRrUY0ZcHfqayckHn9Yu+b/wugY3YMSlwNxunKZ0HS6eM1QrvG5bprQCbwuTrZ0mnIHVKdwiR3oF2zJKvD3sboHW60h2qAPJmKrxlveomsEC+PPCJxCXHO68/lxq3qsLN03TnIpAvXYtGTi9fizYoAvm6vxaNEYF83V+LxohAvixw++d6T0NbRPLle3DKAWE+wsIQyNf5kZ1MYT7CwhDIlwU+CF2vzOJGJF8WeCXo/ytaLRHJlwIjhwW22Cw2m94z0RCRfFlguwO9Z6IhIvnS/4QFORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYORQYOcYIvLnvyBuylXZz5+9Dfd39Ye87Bg19ey2wfVL/gUn1dXED/QJP8SXA8knbOm7OM5nuGl149hHzAx9L320Ocaww+cbV//QRbgsnSzeOhGQjB/bIl3u49y2Z6D88t2d8eyvwcZ+si8fG5ueaa1rW2PgSGoaengBw0eT4q3J5psZrKYMObr8a+QQ4hRsbj5kOXZ4xYnxx3uAXh8S2283t82Kq2dZN169HpRk6sGe+3MO97+YhLU/NNHbgtHHsS/nw4LWftvEl7FkAk4ug+fAVcAhL2/jDBd/hJeAU9vc3JzZ9k9E+a5ndVHzKdMZuXhXSwLcG2DWqydCBPfRlHu59r07o73vS2IFXj4Z6c59p0HIya0yMcxksvSs967xd8vP13Qut5QtDwSlcddTnCJyOiA542t4HfjQV2U0DIjoersF/Hxg6sIe+zMO9b+rk1rXjjB34a5/0Sz8kTrXOqW5aNZQvS++phW/713W8oqvO34TkkIqlo8EpDDB/YltsZGtgnHQ0cgibL92f79yD9/g1GzuwZ77cw73vsyFtbw0xdmDYP7Gff2zphcf9+k4o5MsVM6TVAdaOY5Jjo3PBfUcdglvCZ++25Q+Yk3VfslMYsgeXsa3hxWlg7MCe+XIP977nggeN+Njggb1CEvYAQwf2Cg19DRJY6bywK9jPg7ti8PNgwhjwwI/Z6nSeh7YI5MsDN+yLTRDlj9o7EMjXGfhDS3zGyzpPRUME8uWBI3bUAOToPBUNEciXB7ZXlOzXeSKaIpAvD7ykre05nSeiKQL58sBLK8vRnGN6gkC+PHBxSkqxzhPRFIF8eeAVUVFROk9EUwTy5YGzdJ6F1gjkywOvOfyFIOf9HIF8eeAtGZmZOk9EUwTy5YHXzqt6U+eJaIpAvs636HWQ6snWX+YocKZXJ9grCOTLA7/0Skm8J1tHr14nI+GtXp1gr6DmWxtpsViCctkYgS8PXPbC8jJXW9TOloSn2tk4+uBRGe/cWcIMNd/rGenp6XOz2RiBLw8cFRUWprqZlX0YEIMwQ833Zo7Van2K/yYCge+tT3S4/N0K24PHZbAxAuEOXPrWLUtKSprOz5QR+PLARUUnEl1t0fxRQUHBUhsbIxBmqPn+rBf0rvkKOA5vleVyKh3rjyhwrgdVeeANGzZ91eWG9pp2PrixU/rJMd7aPWGv+b5AAel42Zag8MzF/whwWOkp3e7mX1HzrXstLS1t9rvd8128q1DGtpUAl4dGywm4BJD65PMy4hdJ0/ljmpzXaqWXUIhFRvhy1zPigUNDHw3tfHV2R0xCzC42qk2W3rJCt3ZP2GtcCTf/Zq+ciBKA9avl6//s7kKzmu/NnY5j8M7u+S62y+9gkwJfDJOvPxp+XvLNka/f5/Ad/4GcmQ7frfI7HFLx5YEzATZ2Xi9NClbwYYX0brJ+r7pwzVU5js+1lSrtkOXSbqLwOrSE1KoIT1F4huZ2Q9it77XXpX0l8j1V3x9GB8kZ1dKTgXvKlwd+prJyQef1i75v/C6BjdgxKXA3G6eFy7PMlE4ZL01REA5pA9iksD4oWyXwjlny9WFvA7ROV7pDFUDeTIW3rFfdBP65vtVK05/eKvmGydfP0tWXBz6RuOR45/XnUuNWVbh5mu5cBPKla9HI8e5aNBYE8vXuWjQWBPJ1fy0aIwL5ssDtn+s9DW0RyZfvwSkHhPkIC0MgX+dHdjKF+QgLQyBfFvggdL0yixuRfFnglexKnTiI5EuBkcMCW2wWm03vmWiISL4ssN2B3jPREJF86X/CghwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBwKjBxjBN7cd+QN2Uq7ufP3ob6LNZpO7+KRLKMnjHstsH1S/4FJ9XVxA/0CT/ElwPJJ2zpuzjOZ7hpdePYR8wMfS99tDnGsMPnG1f/0EW47V4b2DyqDZMMG9ki2ZKL/8FxwL1uf9os34MZj/R7+rieMeyvwcZ+si8fG5ueaa1rW2PgSGoaengBw0eT4q3J5psZrKYMObr8a+QQ4nRsbj5kOXZ4xYnxx3uAXh8S2283t82Kq2dbLH6qPizZuYM9kNw9peWomuJf9m3XKG2D99ZUnFxo5cNo49qV8ePDaT9v4EvYsgMlF0Hz4CjicpW384YLv8BJwOvv7mxObvslon7XMbio+ZTpjN68KaeBbvzChPnGYcQN7Jnt1Qn/fk+BeFmDaG7AsBrIeNHLg1aOh3txnGrSczBoT41wGS29MzzpvlxR9ffdCa/nCUHA6Vx31OQKnI6IDnrb3gR9NRXbTgAjn1uceDrCMMG5gz2RTJ7eudbwS3MmywH94At4Zb+TAX/ukX/ohcap1TnXTqqF8WXpPLXzbv67jRV11/iYkh1QsHQ1OZ4D5E9tiI1sD46QDksPZfOn+fL71P7bVzY03bmDPZJ8NaXtrCLiXZYF3B/x77iIjB4b9E/v5x5ZeeNyv74RCvlwxQ1odYO04LDk2Ohfcd9QhuOV89m5b/oA5WfclO50he3AZ2/pfk/vN+JdxA3sqO2jErZ8oVWR33XvXL+6tj7tvWoWhA3uF5OwBxg3sFZ7JMvAEVjo17Arq82BFDH0eTBgDHvgxW53O89AWgXx54IZ9sQmi/FF7BwL5OgN/aInPeFnnqWiIQL48cMSOGoAcnaeiIQL58sD2ipL9Ok9EUwTy5YGXtLU9p/NENEUgXx54aWU5jnNMDxHIlwcuTkkp1nkimiKQLw+8IioqSueJaIpAvjxwls6z0BqBfHngNYe/EOS8nyOQLw+8JSMzU+eJaIpAvjzw2nlVb+o8EU0RyNf5Fr0OUj3Z+sscBc706gR7BYF8eeCXXimJ92Tr6NXrZCS81asT7BXUfGuXJSUl/f6/2RiBLw9c9sLyMldb1M62WCxT7WwcffCojHfuLGGGmm/z+9nZ2Qv5hWoD+n46X4HtrrfngaOiwsJUH9a6j30xoHC3UPO9sVHaTaOtbGxA3/Wr98r4s8pJ/a1PdLj83Qrbg8dlsLEBhbuLa99IyffBLWzcY76Nq9PkrG7sxiOt3yqf0iG3gYuKTiS62qL5o4KCgqU2NsYSWM33Z72gdym9gUqHt4tTPpDzyHnXD9T2uEXO7y51N/CGDZu+6nJDe007H9zYKf3kGK/VW9b3BQpIx8u2BIVnLv5HgMNeHpPc+l7PSE9Pn5vdPd/FuwplbFspBQ6TP9DRcJXAzYHyByqMKulu4NDQR0M7X53dEZMQs4uNapOlnypDt3ZP2GtSn3xeRvwiSfg38kPP3ogS749Jbn2b37darQs/6J7vYrv8Dja1wNZZ8h01bLM0iykKd5jb7cCZABs7r5cmBSv4sKK8vHz9XnXhmqtyHJ9rK1XaIcsB6kIU3oFCaqXAOfJ/Yd+iHhR261v3mnR4nP2uqu+PCxXeNhLbuhG413154GcqKxd0Xr/o+8bvEtiIHZMCd7NxWrg8y8xcgEtTguSESMKbFNYHZasE3qH0in4boHW60h2qAPJmyteHv+om8M/1rVbyfbRV8g2T32GWrr488InEJcc7rz+XGreqws3TdOcikC9di0aOd9eisSCQr3fXorEgkK/7a9EYEciXBW7/XO9paItIvnwPTjkgzEdYGAL5Oj+ykynMR1gYAvmywAeh65VZ3IjkywKvZFfqxEEkXwqMHBbYYrPYbHrPRENE8mWB7Q70nomGiORL/xMW5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5FBg5Pw/6ugBSwerxGAAAAAASUVORK5CYII=\n"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "ragg.func(d, numloc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
